mathics/docpipeline.py

```diff
@@ -699,95 +633,71 @@ def show_and_return():
+    """Print a report with the results of the tests"""
+    test_status = test_pipeline.status
+    test_parameters = test_pipeline.parameters
+    total, failed = test_status.total, test_status.failed
+    builtin_total = test_pipeline.builtin_total
+    skipped = test_status.skipped
+    if test_parameters.max_tests == MAX_TESTS:
+        test_pipeline.print_and_log(
+            f"{total} Tests for {builtin_total} built-in symbols, {total-failed} "
+            f"passed, {failed} failed, {skipped} skipped.",
+        )
+    else:
+        test_pipeline.print_and_log(
+            f"{total} Tests, {total - failed} passed, {failed} failed, {skipped} "
+            "skipped.",
+        )
+    if test_status.failed_sections:
+        if not test_pipeline.parameters.keep_going:
+            test_pipeline.print_and_log(
+                "(not all tests are accounted for due to --)",
+            )
+        test_pipeline.print_and_log("Failed:")
+        for part, chapter, section in sorted(test_status.failed_sections):
+            test_pipeline.print_and_log(f"  - {section} in {part} / {chapter}")
+
+    if test_parameters.generate_output and (
+        test_status.failed == 0 or test_parameters.doc_even_if_error
+    ):
+        save_doctest_data(test_pipeline.output_data)
+        return
 
 
 def test_all(
-    quiet=False,
-    generate_output=True,
-    stop_on_failure=False,
-    start_at=0,
-    max_tests: int = MAX_TESTS,
-    texdatafolder=None,
-    doc_even_if_error=False,
-    excludes: set = set(),
+    test_pipeline: DocTestPipeline,
+    excludes: Optional[Set[str]] = None,
 ) -> int:
     """
     Run all the tests in the documentation.
     """
-    if not quiet:
+    test_parameters = test_pipeline.parameters
+    test_status = test_pipeline.status
+    if not test_parameters.quiet:
         print(f"Testing {version_string}")
 
-    if generate_output:
-        if texdatafolder is None:
-            texdatafolder = osp.dirname(
-                settings.get_doctest_latex_data_path(
-                    should_be_readable=False, create_parent=True
-                )
-            )
-
-    total = failed = skipped = 0
     try:
-        index = 0
-        failed_symbols = set()
-        output_data = {}
-        sub_total, sub_failed, sub_skipped, symbols, index = test_tests(
-            index,
-            quiet=quiet,
-            stop_on_failure=stop_on_failure,
-            start_at=start_at,
-            max_tests=max_tests,
+        test_tests(
+            test_pipeline,
             excludes=excludes,
-            generate_output=generate_output,
-            reload=False,
-            keep_going=not stop_on_failure,
         )
-
-        total += sub_total
-        failed += sub_failed
-        skipped += sub_skipped
-        failed_symbols.update(symbols)
-        builtin_total = len(_builtins)
     except KeyboardInterrupt:
         print("\nAborted.\n")
-        return total
+        return
```

Comment:
We used to return `total` and now we don't. That's okay, but please change the function signature to drop the `int`. Thanks.